### PR TITLE
Annotation @phpstan-latte-var

### PR DESCRIPTION
**Pull request by MartinMystikJonas**

### CODE REVIEW

File: src/Template/Variable.php [hunk 1]
```diff
@@ -44,4 +45,43 @@ public function jsonSerialize()
           'type' => $this->getTypeAsString(),
         ];
     }
+
+    /**
+     * @param array<Variable> ...$variableArrays
+     * @return array<Variable>
+     */
+    public static function union(array ...$variableArrays): array
```
Comment (MartinMystikJonas) [inline src/Template/Variable.php]: This logic was used on multiple places so i extracted it here

File: src/Collector/VariableCollector.php [hunk 2]
```diff
@@ -74,28 +73,38 @@ public function processNode(Node $node, Scope $scope): ?array
         }
 
         if ($nameNode instanceof Expr) {
-            return null;
+            $variableName = null;
```
Comment (MartinMystikJonas) [inline src/Collector/VariableCollector.php:88]: I need to catch assignments that cannot be evaluted but can have annotation for example `$this->template->{$varName}` can be annotated to allow analysis

File: docs/annotations.md [hunk 3]
```diff
@@ -20,7 +24,7 @@ class MyControl extends Control
             /** @phpstan-latte-ignore */
             $this->template->something = $param; // <-- this variable will not be collected (by default all template variables assigned inside method are collected)
         } else {
-            $this->render(__DIR__ . '/MyControl.latte');
+            $this->template->render(__DIR__ . '/MyControl.latte');
```
Comment (MartinMystikJonas) [inline docs/annotations.md:27]: Example code was invalid I fixed it

**Comment by MartinMystikJonas** [comment]
I need to fix tests. It works on local but have some errors on CI.

**Comment by MartinMystikJonas** [comment]
@lulco Tests fixed ready for review

File: src/Collector/Finder/VariableFinder.php [hunk 4]
```diff
@@ -44,11 +54,10 @@ public function __construct(CollectedDataNode $collectedDataNode, TypeSerializer
      */
     public function find(string $className, string $methodName): array
     {
-        return array_merge(
-            $this->collectedVariables[$className][''] ?? [],
-            $this->findInParents($className),
+        return Variable::merge(
```
Comment (lulco) [inline src/Collector/Finder/VariableFinder.php]: This is confusing for me. Why Variable has static merge method where all other variables are somehow merged? I would create some service with non static method or at least helper.

File: src/Collector/Finder/VariableFinder.php [hunk 5]
```diff
@@ -18,24 +19,33 @@
...
-            $methodName = $collectedVariable->getMethodName();
-            if (!isset($this->collectedVariables[$className][$methodName])) {
-                $this->collectedVariables[$className][$methodName] = [];
-            }
-            $this->collectedVariables[$className][$methodName][] = $collectedVariable->getVariable();
+        $assignedVariables = VariableCollector::loadData($collectedDataNode, $typeSerializer, CollectedVariable::class);
+        foreach ($assignedVariables as $variable) {
+            $className = $variable->getClassName();
+            $methodName = $variable->getMethodName();
+            $this->assignedVariables[$className][$methodName] = Variable::union($this->assignedVariables[$className][$methodName] ?? [], [$variable->getVariable()]);
```
Comment (lulco) [inline src/Collector/Finder/VariableFinder.php]: Union on variable should just union actual type of variable with new type. Not create new array of all variables. This should be service / helper outside of Variable

**Review by lulco — CHANGES_REQUESTED** [review]
This PR is huge again, I need bigger display.

File: src/Collector/Finder/VariableFinder.php (hunk 5 above)
Comment (MartinMystikJonas) [reply src/Collector/Finder/VariableFinder.php]: I was thinking about helper too. I will change it.

**Comment by MartinMystikJonas** [comment]
Well most of it are tests :-) There are huge amount of combinations taht needed to be tested.

File: src/Collector/Finder/VariableFinder.php (hunk 5 above)
Comment (MartinMystikJonas) [reply src/Collector/Finder/VariableFinder.php]: Any idea of good name? Simply VariableHelper?

Comment (lulco) [reply src/Collector/Finder/VariableFinder.php]: probably yes :)

**Comment by MartinMystikJonas** [comment]
Updated and rebased on main

**Pull request merged**